lib/tls/session.toit

```diff
@@ -189,14 +192,115 @@ class Session:
+
+  // However when we have a record that doesn't end on a message boundary
+  // we have to switch to a different mode.  We create a synthetic record
+  // and fill it with a single message.
+
+  // At this point we may have a new issue: If we didn't end on a record
+  // boundary, the buffered reader has some amount of data that belongs to
+  // the next fictional record.  We unget 5 bytes of data to create an
+  // artificial record boundary.
+
+  // At some point MbedTLS gets a CHANGE_CIPHER_SPEC_ message, and after
+  // this point the data from the other side is encrypted.  This happens
+  // fairly late in the handshaking, and we have to hope that no more
+  // fragmented packets arrive after this point, because we can no longer
+  // understand the message data and defragment it.
+
+  // Reads and blocks until we have enough data to construct a whole
+  // handshaking message.  May return an artificial record, (defragmented
+  // from several records on the wire).
+  // If the first record is an applicaton data record (handshaking is over) it
+  // returns an arbitrary byte array of data to be passed to MbedTLS, since
+  // MbedTLS can defragment OK once the handshake is over.
+  extract_first_message_ -> ByteArray:
+    if switched_to_encrypted_ or (reader_.byte 0) == APPLICATION_DATA_:
+      return reader_.read
+    header := reader_.read_bytes 5
+    content_type := header[0]
+    message_size / int := ?
+    if content_type == ALERT_:
+      message_size = 2
+    else if content_type == CHANGE_CIPHER_SPEC_:
+      message_size = 1
+      switched_to_encrypted_ = true
+    else:
+      if content_type != HANDSHAKE_:
+        // If we get an unknown record type the probable reason is that
+        // we are not connecting to a real TLS server.  Often we have
+        // accidentally connected to an HTTP server instead.  If the
+        // response looks like ASCII then put it in the error thrown -
+        // it may be helpful.
+        reader_.unget header
+        text_end := 0
+        while text_end < 100 and text_end < reader_.buffered and is_ascii_ (reader_.byte text_end):
+          text_end++
+        throw "Unknown TLS record type: $content_type - server replied '$(reader_.read_string text_end)'"
+      reader_.ensure 4  // 4 byte handshake message header.
+      message_size = (reader_.byte 1) << 16
+      message_size += (reader_.byte 2) << 8
+      message_size += (reader_.byte 3)
+      message_size += 4  // Encoded size does not include the 4 byte handshake header.
+
+    if message_size >= 0x4000: throw "TLS handshake message too large to defragment"
+    // Make an artificial record that was not on the wire.
+    record := ByteArray message_size + 5  // Include space for header.
+    record.replace 0 header
+    // Overwrite record size of header.
+    BIG_ENDIAN.put_uint16 record 3 message_size
+    remaining_in_record := BIG_ENDIAN.uint16 header 3
+    while message_size > 0:
```

Comment:
Feels wrong to call it `message_size` when it is changed. -> `remaining_message_size` ? (not really happy about the name, so if you have a better one...)